Develop

Kvarki-rM opened the pull request.

alozhkin commented:
На будущее: у коммитов должны быть осмысленные названия

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+    String lang = "";
+
+    static ArrayList<String> langDate;
+    static ArrayList<String> FileDate;
```
названия переменных нужно писать с маленькой буквы.

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+
+    static {
+        try {
+            langDate = new ArrayList<>(Files.readAllLines(Paths.get("src\\main\\resources\\language.txt")));
```
Вместо \\\\ стоит использовать /, потому что он работает на Windows И Linux.

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+                            .get();
+                    reCrawl(new HTML(doc, url));
+                } catch (SocketTimeoutException e) {
+                    System.out.println("Превышено время подключения(.timeout(5 * 1000)) " + url);
```
лучше просто написать время

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+                } catch (SocketTimeoutException e) {
+                    System.out.println("Превышено время подключения(.timeout(5 * 1000)) " + url);
+                } catch (SSLProtocolException e) {
+                    System.out.println("SLProtocolException");
```
SSLProtocolException. Лучше не дублировать название исключения, писать на английском и указывать ссылку.

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+            if (act(url, dateList)) {
+                System.out.println(url);
+                count++;
+                dateList.add(url);
```
Лучше печатать ссылку сразу с указанием результата.

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+
+
+    public String lastCharFixer(@NotNull String inURL) {
+        if (inURL.charAt(inURL.length() - 1) == "/".charAt(0))
```
Лучше написать просто '/'

alozhkin commented on src/main/java/util/HTML.java:
```diff
@@ -1,22 +1,34 @@
     private Link url;
+    private String language;
 
-    public HTML(String html, Link url) {
```
У конструкторов класса лучше не менять сигнатуру, ибо их кто-то уже вызвал в своём коде в первоначальном виде. Найти способ перевести String в Document и создать отдельный конструктор с Document

alozhkin commented on src/main/java/crawler/TestCrawler.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class TestCrawler {
```
Тесты должны быть unit тестами: располагаться в папке test/java, быть написаны с помощью библиотеки junit. За примерами в семестровые курсы либо в наш текущий проект.

alozhkin commented on src/main/java/crawler/DefaultLinkFilter.java:
```diff
@@ -1,14 +1,23 @@
+    public List<Link> filter(@NotNull List<Link> links, String domain) {
+        List<Link> out = new ArrayList<>();
+        for (Link link : links) {
+            if (!out.contains(link))
```
Вместо вложенных if лучше использовать &&

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+
+    boolean act(@NotNull Link url, List<Link> dateList) {
+        for (String name : FileDate)
+            if (url.toString().substring(url.toString()
```
Лучше написать в Link отдельный метод, который возвращает расширение файла.

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+        return inURL + "/";
+    }
+
+    boolean act(@NotNull Link url, List<Link> dateList) {
```
Логичнее другой подход. Если нам нужны только определённые файлы, то брать расширение ссылки и сравнивать с ними (белый список), а не составлять список не подходящих расширений (чёрный список).

Nekobitlz replied:
Андрей тебе всё по факту написал, но я от себя ещё добавлю: 
1) Старайся побольше выносить результаты функций в переменные (но без фанатизма офк), чтобы это лучше читалось. Очень много мест, особенно в этом методе, где ты вызываешь url.toString() несколько раз и это не очень красиво, лучше выносить в начале метода в какую-то переменную String domain = url.toString() и потом её использовать, гораздо лучше будет читаться
2) Проблема с неймингом функций, конкретно название этой функции act не дает представление о том что делает функция и плохо читается, когда ты используешь if (act(url, datelist))

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+                return false;
+
+        for (String name : langDate)
+            if (url.toString().contains("/" + name + "/") && !url.toString().contains("/" + lang + "/"))
```
Опять же, лучше написать метод в Link, который вычленяет язык.

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+            if (url.toString().contains("/" + name + "/") && !url.toString().contains("/" + lang + "/"))
+                return false;
+
+        return !url.toString().contains("#") && !(url == domain) &&
```
Те проверки, которые отбрасывают большинство сайтов, нужно проводить первыми.

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+                return false;
+
+        return !url.toString().contains("#") && !(url == domain) &&
+                url.contains(domain) && !dateList.contains(url);
```
Весь этот код должен быть в классе Link в функциях типа: getLanguage(), [isContainsFragmentIdentifier()](https://en.wikipedia.org/wiki/Fragment_identifier) (с комментарием о том, что это такое), getHost() и т.д.

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+        return dateList;
+    }
+
+    private void reCrawl(@NotNull HTML html) {
```
В чём была изначальная идея: Crawler просто отдаёт все ссылки из одного документа и затем LinkFilter отфильровывает не нужное. LinkFilter знает доменное имя, хранит в себе статическую коллекцию, куда сваливает все уже обработанные ссылки. После фильтрации он отдаёт все ссылки в очередь, откуда они снова попадают в crawler. Так что не нужно обходить документ рекурсивно в Crawler. Обработка всего сайта обеспечивается очередью.

Kvarki-rM replied:
теперь я точно запутался в твоих приказах, если мне не надо рекурсивно искать все ссылки с домена , то зачем ты меня парил с подключениями к сайтам. На вход в crawler ты написал html, который уже содержит все ссылки. А фильтр отфильтровывает ненужные ссылки, соответственно мне не нужно работать с подключениям к сайтам

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+import javax.net.ssl.SSLProtocolException;
+
 public class DefaultCrawler implements Crawler {
+    List<Link> dateList = new ArrayList<>();
```
1. data, а не date.
2. Названия переменных не информативны. И без того понятно, что лист содержит некие данные. Они должны чётко обозначать назначение переменной. К примеру в данном случае лучше подошло бы allLinksFromDomain

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+
 public class DefaultCrawler implements Crawler {
+    List<Link> dateList = new ArrayList<>();
+    Link domain = new Link("");
```
Внутренние переменные класса должны быть private

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+
+import javax.net.ssl.SSLProtocolException;
+
 public class DefaultCrawler implements Crawler {
```
По архитектуре, создать конструктор, который принимает необходимые данные, вроде domain

alozhkin commented on src/main/java/crawler/TestCrawler.java:
```diff
@@ -0,0 +1,37 @@
+        String rootURL1 = urlFixer("https://www.zwick-edelstahl.de");//https://hakim.se/404 https://www.zwick-edelstahl.de/en https://www.albrecht-fachbau.de/
+        try {
+            Document document = Jsoup.connect(rootURL1).referrer("http://www.google.com").get();
+            List<Link> zzz = new DefaultCrawler().crawl(new HTML(document, new Link(rootURL1)));
```
Так переменные называть нельзя

alozhkin commented on src/main/java/crawler/TestCrawler.java:
```diff
@@ -0,0 +1,37 @@
+            List<Link> zzz = new DefaultCrawler().crawl(new HTML(document, new Link(rootURL1)));
+            for (Link each : zzz) System.out.println(each);
+            System.out.println();
+        } catch (UnknownHostException e) {
```
Он же ловится внутри самого Crawler

alozhkin commented on src/main/java/crawler/TestCrawler.java:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+
+    public static String urlFixer(String inURL) {
```
Не в тестовом классе

alozhkin commented on src/main/java/crawler/TestCrawler.java:
```diff
@@ -0,0 +1,37 @@
+
+public class TestCrawler {
+
+    public static void main(String[] args) {
```
Тесты должны проверять какие-то факты, а не просто печатать список ссылок. Например можно взять игрушечный сайт с 3 ссылками и убедиться, что crawler их все выдаёт.
+ доджен быть протестирован весь функционал, от # до языков. Каждый в отдельном тесте.

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
 package crawler;
```
По стилю. Я не выпускал рекомендации, к сожалению, но лучше ставить {} даже в однострочных if и for. Исключение, если это if () return или if () continue. + кое-где есть лишние строки.

alozhkin commented:
Только не закрывай пулл реквест, коммить изменения в него

Nekobitlz commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,109 @@
+    boolean act(@NotNull Link url, List<Link> dateList) {
+        for (String name : FileDate)
+            if (url.toString().substring(url.toString()
+                    .length() - 7, url.toString().length() - 1)
```
Магические числа нужно выносить в константы, либо как минимум помечать комментарием

Kvarki-rM committed:
Первая версиия кроулера + неполноценые тесты + филтр

Kvarki-rM committed:
Первая версиия кроулера + неполноценые тесты + филтр

Kvarki-rM committed:
????

Kvarki-rM committed:
????

Kvarki-rM committed:
????

Kvarki-rM committed:
????

Kvarki-rM committed:
????

Kvarki-rM committed:
????

Kvarki-rM committed:
Миннимум фильтра и фильтра ссылок

Kvarki-rM committed:
Миннимум фильтра и фильтра ссылок

Kvarki-rM committed:
Миннимум фильтра и фильтра ссылок

Kvarki-rM committed:
Миннимум фильтра и фильтра ссылок

Kvarki-rM committed:
Недофильтр, кроулер, недотесты

Недоработки:
Фильтр - не обрабатывает разных языков (хз как сделать зная только ссылку или сказывается человеческий фактор к примеру it это может быть и италия, и домен верхнего уровня, и информационные технологии); не обрабатывает ссылки на файлы (тк я не смог прпидумать алгоритм)
Кроулер - выводит все ссылки со страницы, даже те которые идут на файлы, картинки (теоретически этого и не должно тут быть)
Тесты - не на junit тк нечего тестить, но можно проследить некий алгоритм на предюявленных запускатарах вручную
Так же присудствубт каркасы накоторых идей, которые закомичены, что бы не мешались

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,32 @@
+
+        for (Element page : linksOnPage) {
+            Link url = new Link(page.attr("abs:href"));
+            if (!url.toString().equals("")) {//page.text()
```
Закомменченный код нужно убрать. [Мартин, Чистый код, C5, c. 324].

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,32 @@
-        return null;
+    public List<Link> crawl(@NotNull Html html) {
+        List<Link> list = new ArrayList<>();
+        Document doc = Jsoup
```
Раз я начал ревьюить это файл, то: лучше в одну строчку. В строке может быть до 120 символов и этим нужно пользоваться.

alozhkin commented on src/main/java/crawler/DefaultLinkFilter.java:
```diff
@@ -1,14 +1,30 @@
-        return null;
+    public Set<Link> filter(@NotNull List<Link> links, String domain) {
+
+        Set<Link> exit = new HashSet<>();
```
Лучше назвать result. И строчку сверху убрать

alozhkin commented on src/main/java/crawler/DefaultLinkFilter.java:
```diff
@@ -1,14 +1,30 @@
     }
+
+
+   // public static boolean formatChecker(@NotNull String lang, @NotNull Link url) { ///не работает тк человек не робот но должно убирать страницы на других языках
```
Не дописанный код лучше держать в других ветках.

alozhkin commented on src/main/java/utils/Html.java:
```diff
@@ -1,20 +1,29 @@
 import java.util.Objects;
 
 public class Html {
     private String html;
```
Ты же хотел его в Document хранить?

Kvarki-rM replied:
Я все еще хочу, но я понял что это никому не пригодится

Kvarki-rM replied:
тут проблема в том что если изначально при коннекте к сайту не будут его в документ класть то Document будет только место больше занимать  и он будет только место занимать

alozhkin replied:
Если пригодится тебе, то вперёд. Главное не забудь создать новый конструктор

alozhkin replied:
Ок, тогда зальёт изменения от всех и разберёмся

alozhkin commented on src/test/java/CrawlerTest.java:
```diff
@@ -0,0 +1,2 @@
+public class CrawlerTest {
```
Тесты не удовлетворительные, к сожалению. Почитай про JUnit. Тесты не должны вызываться через main и сравниваться выводом в консоль.

Kvarki-rM replied:
сделаю, как только пойму на что должен быть способен фильтр и выбрать подходящий сайт

alozhkin replied:
Простенький сайт с ссылочками можно сделать самому.

alozhkin replied:
Что не понятно в функциональности фильтра?

alozhkin replied:
Нужно отдельно тестировать каждую отдельную фичу

alozhkin replied:
Тесты можешь потом залить, отдельным пуллом

alozhkin commented on src/test/java/TestCrawler.java:
```diff
@@ -0,0 +1,34 @@
+
+    public static void main(String[] args) {
+        try {
+            List<Link> zzz = new DefaultCrawler().crawl(new Html(Jsoup.connect(rootURL).get().toString(), new Link(rootURL)));
```
Названия переменных

alozhkin commented on src/main/java/crawler/DefaultCrawler.java:
```diff
@@ -1,15 +1,32 @@
 package crawler;
 
+import org.jetbrains.annotations.NotNull;
```
Надеюсь, ты не выбросил тот код, он нам пригодится. Когда сольём всй воедино я покажу куда его вставить.

Kvarki-rM replied:
нуууу, он возможно где то есть, у меня уже столько версий при потребности я либо его  вспомню либо открою в недрах пк

alozhkin replied:
В гите есть? Просто ты force push делал

alozhkin commented:
Если это пригодится тебе, то вперёд. Только не забывай создать новый
конструктор, а не менять старый

вс, 22 мар. 2020 г. в 18:32, Kirill <notifications@github.com>:

> *@Kvarki-rM* commented on this pull request.
> ------------------------------
>
> In src/main/java/utils/Html.java
> <https://github.com/alozhkin/OPD-2020/pull/9#discussion_r396106576>:
>
> > @@ -1,20 +1,29 @@
>
>  package utils;
>
>
>
> +import org.jetbrains.annotations.NotNull;
>
> +import org.jsoup.Jsoup;
>
> +
>
>  import java.util.Objects;
>
>
>
>  public class Html {
>
>      private String html;
>
>
> Я все еще хочу, но я понял что это никому не пригодится
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/alozhkin/OPD-2020/pull/9#discussion_r396106576>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKI6TO34IZN6OJA7DYW3PQLRIYVPHANCNFSM4LMTZ7IA>
> .
>


-- 
С уважением,
Ложкин Андрей

Kvarki-rM committed:
Поменял List на Set, удалил тест-дебагер, что бы позже сделать тетсты  исправил пару ошибок указанных Адреем

alozhkin committed:
Merge branch 'develop' into develop
